### ISSUE TITLE
go.dev: API for pkg.go.dev

### ISSUE THREAD

**Issue by rhcarvalho** [golang/go#36785]
Prior to [pkg.go.dev](https://pkg.go.dev), [godoc.org](https://godoc.org) has had a JSON API that can be used to, among other things, discover importers of a given package.

Example: https://api.godoc.org/importers/golang.org/x/net/html

Given that [pkg.go.dev](https://pkg.go.dev) does a much better job at tracking importers thanks to Go Modules and the Module Proxy, it would be nice if the community could get access to a public API similar to that of [godoc.org](https://godoc.org).

**Comment by julieqiu** [comment]
@rhcarvalho - it would be helpful to get a sense of what your current use cases are for api.godoc.org, and feature requests are for an API for pkg.go.dev.

It sounds like getting the importers for a package is one of them. With pkg.go.dev being module aware, what specific information about importers would be useful to surface via an API?

For example:
- importers for a specific version of a package
- only importers of the latest version of a package
- any importer for all versions of a package
- something else?

Additionally, what other information would be useful to you to surface via an API?

/cc @tbpg who has also mentioned wanting an API for pkg.go.dev

**Comment by rhcarvalho** [comment]
Without much thought, having an API to answer the more specific question "what are the importers of a specific version of a package" would make plausible to derive the answer to the other items in your list.

At the moment I consume the godoc.org API and scrape data from pkg.go.dev to answer the question "who uses my package". As far as I can tell the data in the "Imported By" go.dev tab is unrelated to the version of the package I'm currently browsing.

Here are the other endpoints in the GoDoc API: https://github.com/golang/gddo/blob/7365cb292b8bfd13dfe514a18b207b9cc70e6ceb/gddo-server/main.go#L901-L904

- `/search`: takes a `q` parameter. Can be useful to explore known packages/modules.
- `/packages`: attempts to return all known packages, no pagination. Doesn't seem too usable as is.
- `/importers/`: useful, hard to compute without central knowledge.
- `/imports/`: dispensable, easy to compute offline with `go list`.

So if we need a more specific request, here it is:

## api.go.dev MVP

- `/importers/{module-or-package}/@v/<version>`: returns the list of importers of a given module/package at a given version. URL scheme could be tuned to match the Go Proxy specification (`go help goproxy`).
- `/search/`: matches the functionality of https://pkg.go.dev/search?q=hello, returns JSON instead.

**Comment by tbpg** [comment]
Another endpoint idea (standalone or as part of a set of info returned for a module): available versions of a given module.

**Comment by myitcv** [comment]
@tbpg 

> Another endpoint idea (standalone or as part of a set of info returned for a module): available versions of a given module.

Shouldn't people use `cmd/go` for that? Not least because of the fall-through semantics of the go env GOPROXY` variable. Noting https://github.com/golang/go/issues/37367

**Comment by julieqiu** [comment]
What information beyond `https://proxy.golang.org/<module>/@v/list` would be provided by that endpoint?

**Comment by tbpg** [comment]
> Shouldn't people use cmd/go for that? Not least because of the fall-through semantics of the go env GOPROXY` variable. Noting #37367

There are some cases where `cmd/go` might not be available and a normal web API would be helpful. I think it's reasonable to assume pkg.go.dev will only return the modules/versions it knows about.

> What information beyond https://proxy.golang.org/<module>/@v/list would be provided by that endpoint?

None. That works. :)

**Comment by myitcv** [comment]
> There are some cases where cmd/go might not be available and a normal web API would be helpful

I'd be wary of encouraging tool authors to query pkg.go.dev/a proxy directly. Because it could well introduce skew compared to the answers from `cmd/go`.

**Comment by adamdecaf** [comment]
In my use-case I don't want to execute `cmd/go` binaries as I just want to grab some metadata from a dozen modules and organize it on a webpage. If I need local clones of those modules it really complicates this internal tool.

**Comment by myitcv** [comment]
In https://github.com/golang/go/issues/37952 I raised the question of whether module/package license file information could be surfaced in the output of `cmd/go list`. Following a conversation on yesterday's [golang-tools](https://github.com/golang/go/wiki/golang-tools) call, we concluded that doing so would be a bad idea; reasons summarised in https://github.com/golang/go/issues/37952#issuecomment-611527845.

@ianthehat instead suggested exposing the information via a pkg.go.dev API, leveraging the fact that the content and presentation on pkg.go.dev has already jumped through the relevant legal hoops. 

This comment is therefore to explicitly request that we include license file information in the API. Thank you!

**Comment by tiegz** [comment]
👋 Thanks for all the new Go tools like pkg.go.dev, they're super useful. Some input on this topic from the perspective of [Libraries.io](https://libraries.io/):

* +1 to a `/search/` endpoint, or maybe just adding pagination to this url so it's possible to get a list of all packages: `https://index.golang.org/index?since=#{project_window}&limit=2000`. Many popular package repositories offer a full listing like this, e.g. `https://packagist.org/packages/list.json` for Packagist 

* > What information beyond https://proxy.golang.org/<module>/@v/list would be provided by that endpoint?

Currently to get the publish times for each version, you have to make N+1 requests: one for `@v/list` and then `@v/xxx.info` for each version and grab the `"Time"` value for each one. Assuming it's difficult to change the Go proxy spec, an API endpoint that returns all versions with metadata would be really nice.

* > As far as I can tell the data in the "Imported By" go.dev tab is unrelated to the version of the package I'm currently browsing.

Anyone know if there are plans to fix this?

**Comment by jba** [comment]
> * > As far as I can tell the data in the "Imported By" go.dev tab is unrelated to the version of the package I'm currently browsing.
> 
> Anyone know if there are plans to fix this?

No immediate plans. We currently gather that information from import statements in the code, so there is no version information attached. The `go.mod` file doesn't have all the version information we need.

So we understand it's an approximation and we want to fix it, but it's going to take some time.

**Comment by pombredanne** [comment]
@julieqiu my use for an API would be to:
1. search and list packages
2. for a package, access all the information as available for a given package in each of the tabs, such as here https://pkg.go.dev/github.com/gin-gonic/gin?tab=licenses ... basically everything that is available as HTML in the frontend https://github.com/golang/pkgsite/tree/master/internal/frontend should be available in a JSON api

**Comment by pombredanne** [comment]
@julieqiu oh and please do not retire http://api.godoc.org/packages unless there is an alternative!

**Comment by eclipseo** [comment]
>  it would be helpful to get a sense of what your current use cases are for api.godoc.org, and feature requests are for an API for pkg.go.dev.

As a downstream package maintainer for Fedora, I'm also interested in an API. We have our own tool, Anitya, to track package releases, but it was not designed to track GIT commits. And many Go packages still don't publish version. So any information about latest published commit, with info like date of the commit, new dependencies, would be very helpful. I would gather data from the API in Python and compare it to the latest version we have for a given package.

**Comment by pombredanne** [comment]
and BTW my general use case is for https://github.com/nexB/scancode-toolkit and related projects to provide my users with license, origin and dependencies details. And for https://github.com/nexb/vulnerablecode where we provide vulnerabilities details.

**Comment by StevenACoffman** [comment]
My present use case is discovering major version upgrades. Currently, [we have to scrape pkg.go.dev](https://github.com/icholy/gomajor/blob/2ce7ff463e5e9811e511b2b8e761a4558991edf2/internal/latest/latest.go#L38-L63). An alternative to pkg.go.dev API, would be if  there were an "official" way (via `go list` for example) in the future to retrieve the latest major version of a module, we could use it in tools like [go-mod-upgrade](https://github.com/oligot/go-mod-upgrade/) and [icholy/gomajor]((https://github.com/icholy/gomajor/) but that would depend on golang/go#40357 (deprecated notification) and golang/go#40323 (notify about newer major versions). Thanks!

**Comment by icholy** [comment]
@StevenACoffman I've updated gomajor to find newer versions using the module proxy. The performance is acceptable.

**Comment by 0x2b3bfa0** [comment]
My use case is importing modules/packages to another package manager ([Guix](https://guix.gnu.org)) where builds are completely reproducible and monolithic vendoring is strongly discouraged. Having a JSON API would simplify the obtention of metadata like licenses, brief package descriptions, dependencies and hashes.

**Comment by 89z** [comment]
I didnt see it mentioned yet, I found this link to be helpful:

https://github.com/golang/gddo/wiki/API

I hope similar API will be available for new site, or at least maintain the existing API.

**Comment by jba** [comment]
We'll be keeping `api.godoc.org` around for a whie.

**Comment by 89z** [comment]
@jba with all due respect, that comment is pretty vague. I am actively using that API:

https://github.com/89z/sienna/blob/master/go-repos/download.go

so if it goes down, Im just going to end up cURLing this:

https://pkg.go.dev/search?q=Query

and scraping the HTML, as I suspect others will be too. So any further detail would be appreciated.

**Comment by myitcv** [comment]
> `/search/`: matches the functionality of https://pkg.go.dev/search?q=hello, returns JSON instead.

Building on this point, it would be great if a real-time fuzzy search API were exposed. That way command line tools could wrap/use this.

**Comment by restuwahyu13** [comment]
Hello everyone, is there a public API for pkg.go.dev ?, if there is I want to use it for my package library i.e. [gopack-cli](https://github.com/restuwahyu13/gopack-cli), for now I'm using the go doc API, for downloading go package library from gopack-cli .

**Comment by jba** [comment]
@restuwahyu13 , keep using `api.godoc.org`.

**Comment by restuwahyu13** [comment]
But is `api.godoc.org` the same as `pkg.go.dev`  sir @jba ?

**Comment by fzipp** [comment]
@restuwahyu13 `pkg.go.dev` doesn't have an API yet, that's why this issue is open and not closed.

**Comment by guseggert** [comment]
I have similar use cases as others, so I resorted to scraping the UI as well (:nauseated_face:), and put it behind an API/CLI: https://github.com/guseggert/pkggodev-client

**Comment by sig-aarena** [comment]
The `/packages` api endpoint, https://api.godoc.org/packages, is returning ```{"error":{"message":"Internal Server Error"}}``` .

Has this endpoint been migrated?

**Comment by tills13** [comment]
Seems like most api.godoc.org endpoints are erroring, currently?

e.g. https://api.godoc.org/packages, https://api.godoc.org/imports/github.com/goburrow/cache, etc.

I'm particularly interested in the `imports` endpoint. Has this moved somewhere else or is there a mirror I can use?

**Comment by pombredanne** [comment]
This is starting to be really problematic. @jba could someone kindly look into this? This is really looking bad

**Comment by pombredanne** [comment]
@julieqiu do you know what happened?

**Comment by jba** [comment]
Will get on this today.

**Comment by pombredanne** [comment]
@jba you rock... that's very kind of you! :bow:

**Comment by pombredanne** [comment]
Note that only alternative today would be to resort to extensive pseudo-random screen scraping of https://pkg.go.dev/search laced in with some extensive recursive hitting of https://index.golang.org/index to get a complete package picture. It can work if this is what is recommended, but it would feel a bit like it is 1999 all over again. ;)
If anything and if that's the way to go, then at least I could create a common service that does this for everyone?

**Comment by jba** [comment]
api.godoc.org should be back up.

**Comment by pombredanne** [comment]
@jba this is responding but not doing anything for me:
```
$ wget https://api.godoc.org/packages
--2022-02-01 14:33:56--  https://api.godoc.org/packages
Resolving api.godoc.org (api.godoc.org)... 142.250.179.147, 2a00:1450:400e:80f::2013
Connecting to api.godoc.org (api.godoc.org)|142.250.179.147|:443... connected.
HTTP request sent, awaiting response...
```
this stays there hanging waiting, no I/O nor CPU used.

**Comment by pombredanne** [comment]
@jba strike this out. It completed after a while!
Thank you ++ :bow:

**Comment by ZiViZiViZ** [comment]
> Prior to [pkg.go.dev](https://pkg.go.dev), [godoc.org](https://godoc.org) has had a JSON API that can be used to, among other things, discover importers of a given package.

Sorry for the stupid question but where are the documentations for using this "JSON API" that [godoc.org](https://godoc.org) had?

**Comment by srenatus** [comment]
@ZiViZiViZ https://github.com/golang/gddo/wiki/API

**Comment by ZiViZiViZ** [comment]
Thanks but this does not seem to have enough information.  I am looking to be able to get the license for a package.
I was hoping there is a way to do something like that:
```
 $ curl -s https://api.godoc.org/search?q=gopkg.in/yaml.v2@v2.4.0 |jq '.results[].license'
"Apache License 2.0"
```

**Comment by tills13** [comment]
@ZiViZiViZ the way I accomplish this is by making a request to `https://pkg.go.dev/gopkg.in/yaml.v2?tab=licenses` and using an HTML parser to grab the license headers. (replace `gopkg.in/yaml.v2` with whichever package you're looking for the license for)

**Comment by ZiViZiViZ** [comment]
Thank you.  That is what I am doing right now but I thought there might be a better way to do this.

**Comment by Alphare** [comment]
I am part of the [Software Heritage](https://www.softwareheritage.org) initiative to archive all versions of public code and packages.

I'm looking for a way of listing all go packages, planning then to download the source for each release. For example in the Rust ecosystem, `crates.io` has an API that allows us to get all the metadata for a known package and each of its releases, which is very practical.

It appears that there only exists ` https://api.godoc.org/packages` which returns a single 300MB+ JSON file containing only URLs to the repositories of packages (and some `import_count` metric that is always 0), but nothing more. `https://index.golang.org/index` is not looking any different. Am I missing anything worthwhile that would help?

Thanks in advance.

**Comment by seankhliao** [comment]
please keep this issue on topic and refer questions to 
https://github.com/golang/go/wiki/Questions

**Comment by shellscape** [comment]
Rather than hiding what I'd consider a lot of really useful comments by folks (because I landed here for many of the same reasons they did, and there is _extremely little available_ with regard to go package APIs) it would be infinitely more useful to explain to folks how to use the wiki for questions - or specify that tangential and related questions here are not welcome, and folks should seek the answers on the wiki. As it is now, that curt reply seems to imply that we should post questions on the wiki itself, which I doubt is what you're actually after.

The go team has been kicking this can for years, so users are naturally going to have a lot of questions. This reply https://github.com/golang/go/issues/36785#issuecomment-579820081 specifically asks users what features they're after in an API - thus the issue is fair game for feature-related questions. Go packages are the very last major ecosystem package registry to lack an API and that which doesn't make its data readily available. Because of this, people are naturally going to have a myriad of questions about trying to access that data. I'd humbly ask @seankhliao  to be a bit more gracious in your moderation of this issue.

**Comment by golightlyb** [comment]
* Use case: generating static documentation, or running a local documentation server, but wanting to be able to query (with javascript, at the time of the page view by a web browser) the "[Imported by: NNN]" number that appears if instead viewed on pkg.go.dev.

* currently api.godoc.org/importers is returning `{"error":{"message":"Internal Server Error"}}`

**Comment by rsc** [comment]
For what it's worth, we have been running api.godoc.org as an unmonitored service (meaning we don't page people for it), approximately "best effort" although even that may be too generous. We discovered today that it has been down for 30+ days because the disk filled. Given that being down for a month was a non-issue, we've decided to leave it down.

The godoc.org redirects will continue to run (approximately forever), but api.godoc.org will no longer be accessible. Or rather it will continue to be inaccessible.

**Comment by golightlyb** [comment]
@rsc fair enough if this is the decision but the previous comment is mine, right at the start of that 30+ day period, saying it's down, and three people agreeing.

If this is the decision, again, fine - but I submit that it shouldn't me based on people not complaining because to be fair I raised this pretty much *as soon as it happened* and it was never fixed

**Comment by shellscape** [comment]
honestly this reflects pretty poorly on the entire team and the project in general. compared to other ecosystems and their package management, and user experience, the go community leaves a lot to be desired.

**Comment by shellscape** [comment]
Nice. Comment on the actions taken, or lack thereof, and get hidden as off-topic. If the team isn't open to polite and honest criticism, lock the topic.

**Comment by gwatts** [comment]
I use the [Dash](https://kapeli.com/dash) doc viewer on my Mac which i find super useful for building a local index of all sorts of docs, Go included - Just found i couldn't install a new Go docset into it, presumably as it can no longer build an index from `api.godoc.org`.

Would be nice to point the author towards an alternate index if this one isn't returning as the usefulness of Dash is greatly diminished for me without having rapid access to the docs of Go packages i use frequently.

**Comment by AlekSi** [comment]
Dash 6.3.1 now relies on the web scraping. It works, but API would be better.

**Comment by avamsi** [comment]
I was looking for an API to search for Go "commands" -- https://deps.dev/_/search/suggest?system=go&kind=package&q=axl (and then https://api.deps.dev/v3alpha/systems/go/packages/github.com%2Favamsi%2Faxl) kinda works (I don't mind that deps.dev/_/search is hacky and potentially unstable), but pkg.go.dev also annotates the commands as such (see https://pkg.go.dev/search?q=axl, for example), so it's easy to tell them apart from vanilla libraries.

**Comment by sfblackl-intel** [comment]
> Because of this, people are naturally going to have a myriad of questions about trying to access that data. I'd humbly ask @seankhliao to be a bit more gracious in your moderation of this issue.

We are using that...but realized that for Go packages in particular, there is no publish date...which led me to this thread.  My main real ask is that publish date be included for Go packages...otherwise my solution appears to be to call something like this for each package/version (https://proxy.golang.org/<module>/@v/v1.9.0.info).

**Comment by dvob** [comment]
I also used api.godoc.org for my project https://github.com/dvob/go-project-usage/
To obtain the importers I now scrape  https://pkg.go.dev/%s?tab=importedby (see https://github.com/dvob/go-project-usage/blob/72c5625c361b7e81fa2d1a35591c30830dc1164f/main.go#L134).
Unfortunately this is prone to break (e.g if CSS class name changes) and for packages with many importers it only returns 20k packages.

**Comment by mark-pictor-csec** [comment]
Another use for the API would be to check if any versions of a module are retracted. Right now the only way I see to get that info is pretty terrible - I must GET `pkgs.go.dev/path/to/pkg/?tab=versions`,  and search the html for the word `retracted`. If it's there then I get to parse the html and figure out which version(s) are so marked.

People at my company are very thankful that go vulnerabilities usually include affected symbols ([example](https://pkg.go.dev/vuln/GO-2024-2598)), as it allows us to report to customers whether they're actually using a vulnerable function with far better accuracy. 

So while vuln reports are much better in go, for retractions we must parse a web page and hope the html doesn't change. This feels like a real step back.

I hope an API can be prioritized.

**Comment by nabokihms** [comment]
+1 for the feature, I'd like to use pkg.go.dev to find the repo with the source code by the package name. Now I must parse the HTML code to do so.

**Comment by seankhliao** [comment]
I think that's #18387

**Comment by seankhliao** [comment]
re: retractions, parsing `go list -json -retracted` or the go.mod directly with https://pkg.go.dev/golang.org/x/mod@v0.22.0/modfile#File seems to make much more sense than going through pkgsite, it'll even work for private dependencies.

**Comment by xxxserxxx** [comment]
My use case for this is services like [shields.io](https://shields.io/badges/endpoint-badge). pkg.go.dev already parses out and processes a lot of Go-specific information that could be used to generate end-user badges in READMEs. It also produces a single piece of information that is -- AFAIK -- not available elsewhere: the number of packages that import a given package. Additional useful information already being parsed:

- The package version
- The software license
- The number of imports

Without running bespoke services to re-do this data extraction that pkg.go.dev is already doing, the alternative for most users is to parse this information out and then commit this information to some repository; this has long been a source of contention for VCS users -- for example, if committed to the project repo itself, it muddies the VCS history.

pkg.go.dev is already extracting all of this information -- making the metadata that it composes available via an API call (or calls) would allow use in badge creation and potentially other useful analysis tools.

For my money, all I'd ask for is a single API that provides the metadata processed and compiled by pkg.go.dev in a single JSON or even TOML response. shields.io already supports generating shields from API calls that return a number of different formats.

**Comment by josharian** [comment]
Is the status on this issue that it needs a policy decision on whether any API-ish access is allowed, or a case by case discussion?

I re-found this this morning, hoping for JSON output from /search.

If it’s in case-by-case mode, I can file a sub-issue. If it’s a policy question, I guess I just join the waiting queue. :)

**Comment by jba** [comment]
It's a policy/bandwidth/priority thing.

**Comment by josharian** [comment]
Would CLs be reviewed?

**Comment by jba** [comment]
Not at this time, I'm sorry.

**Comment by anistark** [comment]
Is there any update on apis for https://pkg.go.dev/ ?

**Comment by jba** [comment]
@anistark There continue to be no plans for an API.

**Comment by acheong08** [comment]
I'm curious how the Repository is detected in the details. I don't see any references to it in repositories or go.mod. e.g. gopkg.in/natefinch/lumberjack.v2 resolves to github.com/natefinch/lumberjack. There is a mention in the readme but I've also seen some without mentions still getting tagged with the current repository. If it's some internal information derived from private information, I'd like to see an API for that.

**Comment by xplshn** [comment]
I would suggest a way to identify if a package is a program or a library, it would be really helpful

**Comment by jba** [comment]
@xplshn If you mean in the UI, that is a separate issue. If you mean as part of the API, you're in the right place, but there are no plans for it.

**Comment by chisleu** [comment]
Posting my use case in case it inspires others.

I built pypi-scout-mcp https://github.com/chisleu/pypi-scout-mcp

It's a way to inject context into models to help where they suffer. Models are limited to their training data which has old versions and interfaces for libraries and can be problematic for developers.

What I find helpful for the model is to give it the ability and instruction to always try to use the latest compatible versions available. So pypi-scout-mcp can get the package metadata, versions available, and the github repo from pypi. That let's it download the README.md for the model, or download the entire source tree and send it to gemini for interface summarization.

With pkg.go.dev, none of this wasteful work is needed. All the information I need the model to have is right there, already summarized. 

An API would be extremely helpful for this use case.

**Comment by H0llyW00dzZ** [comment]
This would also be beneficial for MCP implementation, as issues in #75857

**Comment by TheBlakkthorn** [comment]
Regarding use cases: I need to flesh out SBOM information that includes package version release dates and latest version w/ release date. I'd rather not scrap pkg.dev.go to automate this process.

**Comment by jba** [comment]
> What I find helpful for the model is to give it the ability and instruction to always try to use the latest compatible versions available. So pypi-scout-mcp can get the package metadata, versions available, and the github repo from pypi. That let's it download the README.md for the model, or download the entire source tree and send it to gemini for interface summarization.

Can you specify exactly the API you would need for this? I'm not exactly sure what you're doing.
It seems you have the package name (import path) for the model. Do you also have the module name? If not, note that the same import path may occur in more than one module. If you have the module name, why can't you use the proxy to list all the versions?

**Comment by jba** [comment]
> I would suggest a way to identify if a package is a program or a library, it would be really helpful

Please provide a use case for this.

**Comment by jba** [comment]
> Regarding use cases: I need to flesh out SBOM information that includes package version release dates and latest version w/ release date. I'd rather not scrap pkg.dev.go to automate this process.

Can't you use the proxy for this?

**Comment by xplshn** [comment]
> > I would suggest a way to identify if a package is a program or a library, it would be really helpful
> 
> Please provide a use case for this.

[This](https://github.com/pkgforge-go/builder) repository which automatically discovers and builds Go projects and programs, it builds on Github CI, and uploads to Github GHCR and proper attestations.

It is currently relying on workarounds due to there not being a proper API

It is a very important repository for the [Soar(rust)](https://github.com/pkgforge/soar) package manager and my package manager ([dbin(go)](https://github.com/xplshn/dbin)), as it provides the entirety  of Go programs accesible via the pkg.go.dev page

**Comment by ethanalee-work** [comment]
Update on this. We are working on introducing a formalized `pkg.go.dev` API: https://github.com/golang/go/issues/76718

**Comment by xplshn** [comment]
> Update on this. We are working on introducing a formalized `pkg.go.dev` API: [#76718](https://github.com/golang/go/issues/76718)

Will there be a field to determine if a package is a library?

**Comment by findleyr** [comment]
Meta comment: since we've opened a separate issue for the implementation (#76718), should we close this issue?

Pragmatically, this issue is about the *existence* of an API, not the design. Furthermore, I'm guessing that there are many people subscribed to this issue that may not want to be spammed with design discussion, and I hope there is much design discussion.

So, perhaps we should close this issue and move discussion of the implementation to #76718. Or we can leave this open (since it's not done), with a redirect to the implementation.

All this to say: @xplshn, you should probably ask your question on the implementation issue.

**Comment by ethanalee-work** [comment]
Thanks Rob, I think for now let's keep this issue open and redirect discussion to https://github.com/golang/go/issues/76718. We will address design queries in that thread.

**Comment by ethanalee-work** [comment]
Hi all,

We've launched the v1beta of the pkgsite API. Please feel free to try it out: [pkg.go.dev/api](https://pkg.go.dev/v1beta/api)!

Accompanying blog post: https://go.dev/blog/pkgsite-api

**Comment by reneleonhardt** [comment]
Great to see a first prototype! I can't find any information in the blog post, what rate limits apply?

**Comment by ethanalee-work** [comment]
> Great to see a first prototype! I can't find any information in the blog post, what rate limits apply?

Hi @reneleonhardt, we are adding that info to the site documentation. The current rate limit is 40 QPS.

**Comment by reneleonhardt** [comment]
Thanks!

It seems there has no batching support been designed for the first prototype to avoid roundtrips and hitting the rate limit.
I.e. instead of a single `/v1beta/package/{path}` to request a body with multiple packages (like a go.mod dependency list) resulting in an array of successful Package responses and an array of Error objects.

Is that something which would be considered for the next prototype?

**Comment by ethanalee-work** [comment]
@reneleonhardt That's certainly something that could be explored. Feel free to create a separate issue for this. I am curious as to what use case this might entail? I think we would need to consider both how this endpoint may be used as well as the potential performance tradeoffs incurred.

**Comment by ethanalee-work** [comment]
Closing this issue out in favour of consolidating feedback through https://github.com/golang/go/issues/76718.

**Comment by jba** [comment]
> It seems there has no batching support been designed for the first prototype to avoid roundtrips and hitting the rate limit. I.e. instead of a single `/v1beta/package/{path}` to request a body with multiple packages (like a go.mod dependency list) resulting in an array of successful Package responses and an array of Error objects.
> 
> Is that something which would be considered for the next prototype?

That would subvert the very purpose of the rate limit. We need to manage load on our DB, and the rate limit accomplishes that. If you could ask for N packages in one request, then you could increase load arbitrarily at the cost of one rate limit token. 

The way to do what you want is to send requests in parallel, and prepare to back off if you hit the limit.
